### PR TITLE
Add init function snippet within service block 

### LINKED ISSUE #34830: Provide a snippet for init function within service block
**Description:**
Currently with in a class block we provide a completion snippet to add a init method as follows. 
Peek 2022-01-25 16-47

However, there is no such support for service definition. Having such a completion item would be an improvement.

### PR DESCRIPTION
**Pull request by TharushiJay**
## Purpose
This PR adds completion support for an init function snippet within the service block.

Fixes #34830

## Samples
Screen-Recording-2022-01-31-at-1

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java [hunk 1]
```diff
@@ -16,10 +16,12 @@
 package org.ballerinalang.langserver.completions.providers.context;
 
 import io.ballerina.compiler.api.symbols.Symbol;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java:19]: Shall we increase the priority of the init function as well. Refer the following where I don't see the snippet at the top
Screenshot 2022-02-07 at 12 32 37

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java [hunk 2]
```diff
@@ -92,4 +97,16 @@ private boolean isClientObject(Node node) {
...
+        if (node.kind() != SyntaxKind.SERVICE_DECLARATION) {
+            return false;
+        }
+
+        ServiceDeclarationNode serviceDeclarationNode = (ServiceDeclarationNode) node;
+        return serviceDeclarationNode.members().stream()
+                .filter(member-> member.kind() == SyntaxKind.OBJECT_METHOD_DEFINITION)
+                .map(member->(FunctionDefinitionNode) member)
+                .noneMatch(funcDef -> "init".equals(funcDef.functionName().text()));
+    }
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java:111]: You can all there together in to a single statement

Comment (nadeeshaan) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java:111]: Also, move the `init` string to a constant/ check whether we have already defined it.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/SnippetGenerator.java [hunk 3]
```diff
@@ -1662,6 +1662,19 @@ public static SnippetBlock getServiceKeywordSnippet() {
...
+     * Get Service Initializer Snippet Block.
+     *
+     * @return {@link SnippetBlock}     Generated Snippet Block
+     */
+    public static SnippetBlock getServiceInitFunctionSnippet() {
+        String snippet = "function init() {" + CommonUtil.LINE_SEPARATOR + "\t${2}" +
+                CommonUtil.LINE_SEPARATOR + "}";
+        return new SnippetBlock(ItemResolverConstants.SERVICE_INITIALIZER_TYPE,
+                generateFilterText(Arrays.asList("init", ItemResolverConstants.FUNCTION)), snippet,
+                ItemResolverConstants.SNIPPET_TYPE, Kind.SNIPPET);
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/SnippetGenerator.java:1675]: We have an already defined init function.  Shall we use that? In the already defined one, we have function arguments added. Let's remove that and modify to reuse here as well.

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java (hunk 1 above)
Comment (TharushiJay) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java:19]: Done 👍 
Screenshot 2022-02-07 at 16 06 51

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java [hunk 4]
```diff
@@ -67,6 +70,9 @@ public ObjectBodiedNodeContextProvider(Class<T> attachmentPoint) {
             completionItems.add(new SnippetCompletionItem(context, Snippet.KW_RESOURCE.get()));
             completionItems.add(new SnippetCompletionItem(context, Snippet.DEF_REMOTE_FUNCTION.get()));
             completionItems.add(new SnippetCompletionItem(context, Snippet.DEF_RESOURCE_FUNCTION_SIGNATURE.get()));
+            if (this.onSuggestInitFunction(node)) {
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java:73]: Shall we rename this to `onSuggestInitMethod(node)` as we are trying to suggest a `method`

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/SnippetGenerator.java [hunk 5]
```diff
@@ -1569,7 +1569,7 @@ public static SnippetBlock getRemoteMethodDeclSnippet() {
      * @return {@link SnippetBlock}     Generated Snippet Block
      */
     public static SnippetBlock getInitFunctionSnippet() {
-        String snippet = "function init(${1:any arg}) {" + CommonUtil.LINE_SEPARATOR + "\t${2}" +
+        String snippet = "function init() {" + CommonUtil.LINE_SEPARATOR + "\t${2}" +
```
Comment (malinthar) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/util/SnippetGenerator.java:1572]: Shouldn't we provide tab navigation for the user to define parameters  in the case of a class definition?

**Review by malinthar — COMMENTED** [review]
Changes look good! I have some minor requests

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java [hunk 6]
```diff
@@ -92,4 +98,15 @@ private boolean isClientObject(Node node) {
                 && ((ObjectConstructorExpressionNode) node).objectTypeQualifiers().stream()
                 .anyMatch(token -> token.kind() == SyntaxKind.CLIENT_KEYWORD));
     }
+
+    public static boolean onSuggestInitMethod(Node node) {
+        if (node.kind() != SyntaxKind.SERVICE_DECLARATION) {
+            return false;
+        }
+
+        return ((ServiceDeclarationNode) node).members().stream()
```
Comment (nadeeshaan) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/ObjectBodiedNodeContextProvider.java:107]: Shall we merge the if check with the return statement?

**Review by nadeeshaan — CHANGES_REQUESTED** [review]

**Review by malinthar — APPROVED** [review]
LGTM

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Comment by mohanvive** [comment]
@TharushiJay Shall we send this to patch branch (not required for the master)

**Review by nadeeshaan — APPROVED** [review]
LGTM

**Review by mohanvive — APPROVED** [review]

**Pull request merged**